lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp

```diff
@@ -847,47 +847,114 @@ struct AIEObjectFifoStatefulTransformPass
       lcm = i * lcm / std::gcd(i, lcm);
     return lcm;
   }
-
   // Function that unrolls for-loops that contain objectFifo operations.
   LogicalResult unrollForLoops(DeviceOp &device, OpBuilder &builder,
                                std::set<TileOp> objectFifoTiles) {
     for (auto coreOp : device.getOps<CoreOp>()) {
       if (objectFifoTiles.count(coreOp.getTileOp()) > 0) {
+        std::vector<scf::ForOp> unrolledLoops;
+        std::map<Operation *, bool> foundMap;
+        std::map<Operation *, int64_t> remainderMap;
+        std::map<Operation *, int64_t> tripCountMap;
         WalkResult res = coreOp.walk([&](scf::ForOp forLoop) {
           // look for operations on objectFifos
           // when multiple fifos in same loop, must use the smallest
           // common multiplier as the unroll factor
-          bool found = false;
+          foundMap[forLoop.getOperation()] = false;
           std::set<int> objFifoSizes;
           Block *body = forLoop.getBody();
-
+          remainderMap[forLoop.getOperation()] = 0;
           for (auto acqOp : body->getOps<ObjectFifoAcquireOp>()) {
             if (acqOp.getOperation()->getParentOp() == forLoop) {
-              found = true;
+              foundMap[forLoop.getOperation()] = true;
               ObjectFifoCreateOp op = acqOp.getObjectFifo();
               objFifoSizes.insert(op.size());
             }
           }
-
-          if (found) {
-            int unrollFactor =
-                computeLCM(objFifoSizes); // also counts original loop body
-            // if loop iterations < unrollFactor, unroll the loop fully
-            if (forLoop.getSingleLowerBound() &&
-                forLoop.getSingleUpperBound() && forLoop.getSingleStep()) {
-              int64_t tripCount =
-                  constantTripCount(*(forLoop.getSingleLowerBound()),
-                                    *(forLoop.getSingleUpperBound()),
-                                    *(forLoop.getSingleStep()))
-                      .value_or(0);
-              if (tripCount < unrollFactor)
-                unrollFactor = tripCount;
-            }
-            if (failed(mlir::loopUnrollByFactor(forLoop, unrollFactor))) {
-              forLoop.emitOpError()
-                  << "could not be unrolled with unrollFactor: " << unrollFactor
-                  << "\n";
-              return WalkResult::interrupt();
+          // If the loop doesn't have acquire and release locks
+          // Push it to the unrolledLoops to avoid unrolling
+          if (!foundMap[forLoop.getOperation()]) {
+            unrolledLoops.push_back(forLoop);
+          } else {
```

Comment:
Can avoid the long `else` and extra indentation with an early exit:
```suggestion
          if (!foundMap[forLoop.getOperation()]) {
            unrolledLoops.push_back(forLoop);
            return WalkResult::advance();
          }
```